lang/src/org/partiql/lang/eval/visitors/StaticTypeVisitorTransform.kt

```diff
@@ -0,0 +1,532 @@
+            return from
+        }
+
+        override fun transformFromSourceJoin(node: PartiqlAst.FromSource.Join): PartiqlAst.FromSource {
+            // this happens before FromSourceScan or FromSourceUnpivot gets hit
+            val outermostJoin = !containsJoin
+            containsJoin = true
+
+            return super.transformFromSourceJoin(node)
+                .also {
+                    if (outermostJoin) {
+                        fromVisited = true
+                        singleFromSourceName = null
+                    }
+                }
+        }
+
+        override fun transformFromSourceScan_expr(node: PartiqlAst.FromSource.Scan): PartiqlAst.Expr {
+            this.scopeOrder = ScopeSearchOrder.GLOBALS_THEN_LEXICAL
+            return transformExpr(node.expr).also {
+                this.scopeOrder = ScopeSearchOrder.LEXICAL
+            }
+        }
+
+        override fun transformFromSourceUnpivot_expr(node: PartiqlAst.FromSource.Unpivot): PartiqlAst.Expr {
+            this.scopeOrder = ScopeSearchOrder.GLOBALS_THEN_LEXICAL
+            return transformExpr(node.expr).also {
+                this.scopeOrder = ScopeSearchOrder.LEXICAL
+            }
+        }
+
+        override fun transformGroupBy(node: PartiqlAst.GroupBy): PartiqlAst.GroupBy {
+            errUnimplementedFeature("GROUP BY")
+        }
+
+        private fun innerTransformExprSelect(node: PartiqlAst.Expr.Select): PartiqlAst.Expr {
+            val new_from = transformExprSelect_from(node)
+            val new_fromLet = transformExprSelect_fromLet(node)
+            val new_where = transformExprSelect_where(node)
+            val new_group = transformExprSelect_group(node)
+            val new_having = transformExprSelect_having(node)
+            val new_setq = transformExprSelect_setq(node)
+            val new_project = transformExprSelect_project(node)
+            val new_limit = transformExprSelect_limit(node)
+            val new_metas = transformExprSelect_metas(node)
+            return PartiqlAst.build {
+                PartiqlAst.Expr.Select(
+                    setq = new_setq,
+                    project = new_project,
+                    from = new_from,
+                    fromLet = new_fromLet,
+                    where = new_where,
+                    group = new_group,
+                    having = new_having,
+                    limit = new_limit,
+                    metas = new_metas
+                )
+            }
+        }
+
+        private fun innerTransformDataManipulation(node: PartiqlAst.Statement.Dml): PartiqlAst.Statement {
+            val from = node.from?.let { transformFromSource(it) }
+            val where = node.where?.let { transformStatementDml_where(node) }
+            val dmlOperation = transformDmlOp(node.operation)
+            val metas = transformMetas(node.metas)
+
+            return PartiqlAst.build {
+                dml(dmlOperation, from, where, metas)
+            }
+        }
```

Comment:
Similar to the other comment regarding `transformExprSelect`. Tried changing this to
```kotlin
return super.transformStatementDml(node)
```
but the order affects the end result, causing some tests for fail.

The original rewriter code called [`innerRewriteDataManipulation `](https://github.com/partiql/partiql-lang-kotlin/blob/master/lang/src/org/partiql/lang/ast/passes/AstRewriterBase.kt#L334) which transforms `dmlOperation` after transforming `WHERE`. The PIG code transforms the operation first. Should the PIG version follow the original rewriter transform order?

Reply:
- Add `kdoc` that says this is needed to adjust the transformation order.
- The `inner` prefix I came up with this is not great.  Can you think of a better name for this function?
- Consider moving this to an abstract class that this `StaticTypeRewriter` derives from.  I know for a fact that this will be helpful in the near future.  (The same abstract class as `innerRewriteSelect`?)

Reply:
- Added `kdoc`
- Changed name to `transformDataManipulationEvaluationOrder`
- Moved to an abstract class